ios/RNMessageStream.m

```diff
@@ -40,14 +42,52 @@ - (instancetype)initWithDisplayInAppNotifications:(BOOL)displayNotifications {
 }
 
 - (BOOL)shouldPresentInAppNotificationForMessage:(MARMessage *)message {
-    NSMutableDictionary *payload = [NSMutableDictionary dictionaryWithDictionary:[message dictionary]];
+    if (self.defaultInAppNotification) {
+        return YES;
+    }
+    __block BOOL result = YES;
+
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        result = [self emitWithTimeout:message];
+
+        dispatch_semaphore_signal(semaphore);
+    });
+
+    dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
+
+    return result;
+}
+
+- (BOOL)emitWithTimeout:(MARMessage *)message {
+    __block BOOL success = YES;
 
-    if ([message attributes]) {
-        [payload setObject:[message attributes] forKey:@"attributes"];
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        NSMutableDictionary *payload = [NSMutableDictionary dictionaryWithDictionary:[message dictionary]];
+
+        if ([message attributes]) {
+            [payload setObject:[message attributes] forKey:@"attributes"];
+        }
+        [self sendEventWithName:@"inappnotification" body:payload];
+        
+        success = NO;
+        dispatch_semaphore_signal(self.eventSemaphore);
+    });
+
+    dispatch_semaphore_wait(self.eventSemaphore, dispatch_time(DISPATCH_TIME_NOW, 5 * NSEC_PER_SEC));
+
+    return success;
+}
+
+RCT_EXPORT_METHOD(notifyInAppHandled:(BOOL)handled) {
+    if (handled) {
```

Comment:
The problem with the current setup is that if the RN side returns NO, it'll wait the full 5 seconds before timing out and returning. Unfortunately Obj-C doesn't have a nice tidy Channel implementation to take care of this like on Android...
I think the reasonable path forward here is to rename `success` to something more specific and add it as a property. It can then be set to the value returned `notifyInAppHandled` before signalling the semaphore (make sure to wrap the BOOL access in `@synchronized`)